sql-plugin/src/main/scala/com/nvidia/spark/rapids/CostBasedOptimizer.scala

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, AttributeReference, Expression}
+import org.apache.spark.sql.execution.{ProjectExec, SparkPlan}
+import org.apache.spark.sql.execution.exchange.ShuffleExchangeExec
+import org.apache.spark.sql.types.DataTypes
+
+object CostBasedOptimizer {
+
+  // the intention is to make the cost model pluggable since we are probably going to need to
+  // experiment a fair bit with this part
+  private val costModel = new DefaultCostModel
+
+  /**
+   * Walk the plan and determine CPU and GPU costs for each operator and then make decisions
+   * about whether operators should run on CPU or GPU.
+   */
+  def optimize(plan: SparkPlanMeta[_]): (Double, Double) = {
+
+    // get the CPU and GPU cost of the child plan(s)
+    val childCosts = plan.childPlans.map(optimize)
+    val childCpuCosts = childCosts.map(_._1)
+    val childGpuCosts = childCosts.map(_._2)
+
+    // get the CPU and GPU cost of this operator
+    val (operatorCpuCost, operatorGpuCost) = costModel.applyCost(plan)
+
+    // calculate total (this operator + children)
+    val totalCpuCost = operatorCpuCost + childCpuCosts.sum
+    val totalGpuCost = operatorGpuCost + childGpuCosts.sum
```

Comment:
My main concern here is that this looks like it is written really only for AQE, but that is not documented anywhere.

The total cost looks like the total cost for the entire query (which would produce incorrect results without AQE splitting the query up into smaller parts)

The ShuffleExchangeExec being an explicit cutoff point, and if the cost is not better for the GPU compared to the CPU then everything in the entire query (or at least the part that we are looking at for AQE) will then be put onto the CPU. I'm not sure that is what we want, especially in the context of some parts of the plan having to be on the CPU for what ever reason.

Reply:
I've removed the shuffle-specific logic for now and will add it back once I'm farther along with testing that part.

Reply:
Do we have a notion for situations as follows:  an operator's 
- GPU cost is say 1 sec, 
- CPU cost is 10 sec 

whereas the wait time for GPU is 20sec, and so to reduce the user-perceived latency we should still utilize CPU?

Reply:
> Do we have a notion for situations as follows: an operator's
> 
>     * GPU cost is say 1 sec,
> 
>     * CPU cost is 10 sec
> 
> 
> whereas the wait time for GPU is 20sec, and so to reduce the user-perceived latency we should still utilize CPU?

No, nothing that advanced yet. I imagine we would do this as a follow on issue once we have a basic mechanism in place.